Add Controller Revision (Implementation of KEP #238)

Edwinhr716 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->


#### What this PR does / why we need it
This PR adds controller revision. This allows the controller to store previous iterations of the LWS object, which then makes it possible to select the correct pod template spec to use if a replica is restarted during rolling update.

#### Which issue(s) this PR fixes
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #238
Fixes #239
Fixes #240 
Fixes #281 
Fixes #291


#### Special notes for your reviewer

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note

```

ahg-g commented on pkg/controllers/leaderworkerset_controller.go:
```diff
@@ -277,8 +300,32 @@ func (r *LeaderWorkerSetReconciler) rollingUpdateParameters(ctx context.Context,
+	// https://github.com/kubernetes-sigs/lws/issues/281
+	// If a leader sts already exists, but the template has not been updated, the templateHash of the leader is
+	// used to keep consistency in cases where two different templateHashes are calculated from the same LWS object
+	stsExists, sts, err := stsCreated(ctx, r.Client, lws)
+	if err != nil {
+		return err
+	}
+	templateHash := utils.LeaderWorkerTemplateHash(lws)
+	if stsExists {
+		templateUpdated, err := templateUpdated(ctx, r.Client, sts, lws)
+		if err != nil {
+			return err
+		}
+		if !templateUpdated {
+			templateHash = sts.Labels[leaderworkerset.TemplateRevisionHashKey]
+		}
+	}
```
This should be called in `Reconcile` since multiple functions need to know the templateHash and whether the lws was updated.

ahg-g commented on pkg/controllers/leaderworkerset_controller.go:
```diff
@@ -229,7 +248,11 @@ func (r *LeaderWorkerSetReconciler) rollingUpdateParameters(ctx context.Context,
 	// Case 2:
 	// Indicates a new rolling update here.
-	if templateUpdated(sts, lws) {
+	hasTemplateUdated, err := templateUpdated(ctx, r.Client, sts, lws)
```
Call this in Reconcile and pass to this function a boolean indicating whether or not an update was observed.

ahg-g commented on pkg/utils/controller/controller_utils.go:
```diff
@@ -61,3 +72,213 @@ func CreateHeadlessServiceIfNotExists(ctx context.Context, k8sClient client.Clie
+// previous version. If the returned error is nil the patch is valid. The current state that we save is the
+// leaderWorkerTemplate. We can modify this later to encompass more state (or less) and remain compatible with previously
+// recorded patches.
+
```
nit: remove empty line

ahg-g commented on pkg/utils/controller/controller_utils.go:
```diff
@@ -61,3 +72,213 @@ func CreateHeadlessServiceIfNotExists(ctx context.Context, k8sClient client.Clie
+	if err != nil {
+		return err
+	}
+	currentRevision, err := GetLeaderWorkerSetRevisionFromTemplateHash(ctx, k8sClient, lws, templateHash)
```
No need for this, truncate all revision that has a different templateHash, no need to get it and have the condition on the name as line 277 does.

ahg-g commented on pkg/controllers/leaderworkerset_controller.go:
```diff
@@ -191,20 +194,36 @@ func SetupIndexes(indexer client.FieldIndexer) error {
+
+	if !existingControllerRevisions {
+		// Updating from version that did not support Controller Revision. Need to create one first before checking if template has been updated
+		log.V(2).Info(fmt.Sprintf("Creating new controller revision create/update operation for %+v ", lws))
```
we can't log the whole lws object

ahg-g reviewed (commented):
First round, I didn't look at tests.

ahg-g commented on pkg/utils/controller/controller_utils.go:
```diff
@@ -61,3 +72,213 @@ func CreateHeadlessServiceIfNotExists(ctx context.Context, k8sClient client.Clie
+	err = json.Unmarshal(str.Bytes(), &raw)
+	objCopy := make(map[string]interface{})
+	specCopy := make(map[string]interface{})
+	spec := raw["spec"].(map[string]interface{})
```
This is saving the whole lws spec not just leaderWorkerTemplate as the comment at the top of the function says. 

We need to save two things only: `LeaderWorkerTemplate` and `NetworkConfig`

Edwinhr716 replied:
Changed to only save those

ahg-g commented on pkg/utils/controller/controller_utils.go:
```diff
@@ -61,3 +72,213 @@ func CreateHeadlessServiceIfNotExists(ctx context.Context, k8sClient client.Clie
+		return err
+	}
+	revisionCount := len(revisions)
+	history.SortControllerRevisions(revisions)
```
No need to sort, I would simplify and remove the sort functions, and just iterate here and find the max revision in the list. If the list is 0, use revision=1

Edwinhr716 replied:
Changed logic of `NextRevision`

ahg-g commented on pkg/utils/controller/controller_utils.go:
```diff
@@ -61,3 +72,213 @@ func CreateHeadlessServiceIfNotExists(ctx context.Context, k8sClient client.Clie
+		return err
+	}
+
+	equalRevisions := history.FindEqualRevisions(revisions, currentRevision)
```
no need for this block of code (172 to 190), just call `CreateControllerRevision`

Edwinhr716 replied:
removed, and removed `UpdateControllerRevision` as well

ahg-g commented on pkg/history/controller_history.go:
```diff
@@ -0,0 +1,356 @@
+
+func (rh *realHistory) CreateControllerRevision(parent metav1.Object, revision *appsv1.ControllerRevision) (*appsv1.ControllerRevision, error) {
+	// Clone the input
+	clone := revision.DeepCopy()
+
+	hash := HashControllerRevision(revision)
+	// Update the revisions name
+	clone.Name = ControllerRevisionName(parent.GetName(), hash)
```
No need for the clone and setting the name here; setting the name is is already done by `NewControllerRevision`

ahg-g reviewed (commented):
more comments on simplifying the history pkg

Edwinhr716 commented on api/leaderworkerset/v1/leaderworkerset_types.go:
```diff
@@ -303,6 +303,20 @@ type LeaderWorkerSetStatus struct {
 	HPAPodSelector string `json:"hpaPodSelector,omitempty"`
+
+	// currentRevision, if not empty, indicates the version of the  worker StatefulSet
+	// used to generate the worker pods in sequence [0,currentReplicas)
```
No longer add status fields

Edwinhr716 commented on api/leaderworkerset/v1/leaderworkerset_types.go:
```diff
@@ -303,6 +303,20 @@ type LeaderWorkerSetStatus struct {
+	CurrentRevision string `json:"currentRevision,omitempty"`
+
+	// updateRevision, if not empty, indicates the version of the  worker StatefulSet
+	// used to generate the worker pods in sequence [replicas-updatedReplicas,replicas)
```
No longer add status fields

Edwinhr716 commented on pkg/utils/controller/controller_utils.go:
```diff
@@ -61,3 +71,189 @@ func CreateHeadlessServiceIfNotExists(ctx context.Context, k8sClient client.Clie
+	}
+	for k, v := range lws.Spec.LeaderWorkerTemplate.WorkerTemplate.Labels {
+		combinedLabels[k] = v
+	}
```
Removed, only add `templateHash` to the labels

Edwinhr716 commented on pkg/controllers/leaderworkerset_controller.go:
```diff
@@ -96,6 +99,16 @@ func (r *LeaderWorkerSetReconciler) Reconcile(ctx context.Context, req ctrl.Requ
+	lws.Status.CurrentRevision = currentRevision.Name
+	lws.Status.UpdateRevision = updateRevision.Name
+	lws.Status.CollisionCount = new(int32)
+	lws.Status.CollisionCount = &collisionCount
```
Changed logic

Edwinhr716 commented on pkg/utils/controller/controller_utils.go:
```diff
@@ -61,3 +71,189 @@ func CreateHeadlessServiceIfNotExists(ctx context.Context, k8sClient client.Clie
+}
+
+// TruncateHistory cleans up all other controller revisions expect the currentRevision and updateRevision
+func TruncateHistory(history history.Interface, revisions []*appsv1.ControllerRevision, updateRevision *appsv1.ControllerRevision, currentRevision *appsv1.ControllerRevision) error {
```
Added truncate history after the rolling update is done

ahg-g commented on test/testutils/util.go:
```diff
@@ -540,6 +544,33 @@ func SetLeaderPodsToReady(ctx context.Context, k8sClient client.Client, lws *lea
 	}, Timeout, Interval).Should(gomega.Succeed())
 }
 
+func CreateControllerRevisionForHashCollision(ctx context.Context, k8sClient client.Client, lws *leaderworkerset.LeaderWorkerSet) {
```
Do we need this?

Edwinhr716 replied:
No, removed

ahg-g commented on test/testutils/wrappers.go:
```diff
@@ -288,3 +293,24 @@ func MakeLeaderPodSpecWithTPUResource() corev1.PodSpec {
 	}
 }
+
+func RawLWSTemplate(lws *leaderworkerset.LeaderWorkerSet) runtime.RawExtension {
```
I am not sure we should be using a function like this one in our tests; our tests shouldn't rely on a copy the function that we partially try to test

Edwinhr716 replied:
In this case it is not needed, but there are cases where we do need to duplicate functions as otherwise a dependency cycle is created.

ahg-g replied:
remove the commented out code please.

Edwinhr716 commented on pkg/utils/controller/controller_utils.go:
```diff
@@ -61,3 +71,189 @@ func CreateHeadlessServiceIfNotExists(ctx context.Context, k8sClient client.Clie
 	}
 	return nil
 }
+
```
Moved, and added comment

Edwinhr716 commented on pkg/history/controller_history.go:
```diff
@@ -0,0 +1,356 @@
+
+// FindEqualRevisions returns all ControllerRevisions in revisions that are equal to needle using EqualRevision as the
+// equality test. The returned slice preserves the order of revisions.
+func FindEqualRevisions(revisions []*appsv1.ControllerRevision, needle *appsv1.ControllerRevision) []*appsv1.ControllerRevision {
```
We don't, now that the logic of `CreateLeaderWorkerSetRevision` has been simplified. Removed.

Edwinhr716 commented on pkg/history/controller_history.go:
```diff
@@ -0,0 +1,356 @@
+	UID             types.UID               `json:"uid"`
+}
+
+func (rh *realHistory) AdoptControllerRevision(parent metav1.Object, parentKind schema.GroupVersionKind, revision *appsv1.ControllerRevision) (*appsv1.ControllerRevision, error) {
```
Without the interface, no

Edwinhr716 commented on pkg/history/controller_history.go:
```diff
@@ -0,0 +1,356 @@
+	return revision, err
+}
+
+func (rh *realHistory) ReleaseControllerRevision(parent metav1.Object, revision *appsv1.ControllerRevision) (*appsv1.ControllerRevision, error) {
```
No. Also, `UpdateControllerRevision` is no longer needed after simplifying the logic of `CreateLeaderWorkerSetRevision`, removed that as well

Edwinhr716 commented on pkg/history/controller_history.go:
```diff
@@ -0,0 +1,356 @@
+	br[i], br[j] = br[j], br[i]
+}
+
+// Interface provides an interface allowing for management of a Controller's history as realized by recorded
```
Removed interface

Edwinhr716 commented on pkg/utils/controller/controller_utils.go:
```diff
@@ -61,3 +72,213 @@ func CreateHeadlessServiceIfNotExists(ctx context.Context, k8sClient client.Clie
+	}
+	for key, value := range lws.Annotations {
+		cr.ObjectMeta.Annotations[key] = value
+	}
```
No need, removed

Edwinhr716 commented on pkg/utils/controller/controller_utils.go:
```diff
@@ -61,3 +72,213 @@ func CreateHeadlessServiceIfNotExists(ctx context.Context, k8sClient client.Clie
+
+func getPatch(lws *leaderworkerset.LeaderWorkerSet) ([]byte, error) {
+	str := &bytes.Buffer{}
+	clone := lws.DeepCopy()
```
No need, removed

ahg-g commented:
Great, i will look later today, please ensure we have sufficient integration test coverage

ahg-g commented on pkg/controllers/leaderworkerset_controller.go:
```diff
@@ -523,8 +549,72 @@ func (r *LeaderWorkerSetReconciler) iterateReplicas(ctx context.Context, lws *le
+	if err != nil {
+		return false, err
+	}
+	return !revisionutils.EqualLeaderWorkerTemplates(baselineLws, lws), nil
```
We should use `EqualRevision` here instead. We compare the existing revision (one fetched in line 591) with one we create in memory from the current lws using `NewRevision`.

We don't want another equality function, `EqualLeaderWorkerTemplates` should not exist.

ahg-g commented on pkg/utils/revision/revision_utils.go:
```diff
@@ -0,0 +1,249 @@
+	return hex.EncodeToString(h.Sum(nil))
+}
+
+func LeaderWorkerTemplateHash(lws *leaderworkerset.LeaderWorkerSet) string {
```
we don't need this anymore, we have a hash function in the history pkg now that we should converge on.

Edwinhr716 replied:
I'm guessing you are referring to this one https://github.com/kubernetes-sigs/lws/pull/277/files#diff-356fe1c7888b3417da9a1bd377ba0eb471c7b7e97c3684a7e88165e6c14a2f14R88-R97? 

I already commented this somewhere else, but I think it also fits here: 

Don't we still need a hashing function that treats NetworkConfig == nil and NetworkConfig.SubdomainShared as the same though?

When updating from a version that doesn't have networkConfig the steps would be:
Create a controller revision where the NetworkConfig si nil. When checking if LWS was updated, it will determine that it isn't because the hash is the same.

Another change is made, similar to https://github.com/kubernetes-sigs/lws/issues/240#issuecomment-2452679190 triggering the lws webhook, and defaulting NetworkConfig. A new controller revision will be created, as the data will now be different. Now, the hash won't match the hash that is in the leaderSts, triggering a rolling update when it isn't supposed to.

ahg-g commented on pkg/utils/revision/revision_utils.go:
```diff
@@ -0,0 +1,249 @@
+	return nil
+}
+
+func EqualLeaderWorkerTemplates(lhs *leaderworkerset.LeaderWorkerSet, rhs *leaderworkerset.LeaderWorkerSet) bool {
```
We don't need this function

ahg-g commented on pkg/controllers/leaderworkerset_controller.go:
```diff
@@ -96,13 +100,31 @@ func (r *LeaderWorkerSetReconciler) Reconcile(ctx context.Context, req ctrl.Requ
+		return ctrl.Result{}, err
+	}
+
+	templateHash := getLeaderWorkerTemplateHash(leaderSts, lws, lwsUpdated)
```
we should not create a template hash here, it should come from the revision returned as suggested in line 109 above

Edwinhr716 replied:
Don't we still need a hashing function that treats `NetworkConfig == nil` and `NetworkConfig.SubdomainShared` as the same though? 

When updating from a version that doesn't have networkConfig the steps would be:
Create a controller revision where the NetworkConfig si nil. When checking if LWS was updated, it will determine that it isn't because the hash is the same. 

Another change is made, similar to https://github.com/kubernetes-sigs/lws/issues/240#issuecomment-2452679190 triggering the lws webhook, and defaulting NetworkConfig. A new controller revision will be created, as the data will now be different. Now, the hash won't match the hash that is in the leaderSts, triggering a rolling update when it isn't supposed to.

ahg-g replied:
We are not comparing hashes anymore though, the hash is just a key to the controller revision. We depend on semantic diff of the revisions.

ahg-g commented on pkg/controllers/leaderworkerset_controller.go:
```diff
@@ -274,11 +293,16 @@ func (r *LeaderWorkerSetReconciler) rollingUpdateParameters(ctx context.Context,
+	if err := revisionutils.CreateLeaderWorkerSetRevision(ctx, r.Client, lws, templateHash); err != nil {
+		log.Error(err, "Creating LWS Revision")
+		return err
+	}
```
We shouldn't be creating here, the creation will be done at the beginning of the reconcile function via `createControllerRevisionIfNonExist`

Edwinhr716 replied:
Removing this one + removing `templateHash` requires changes to the way we set the `templateHash` label in the revision. With these changes, we rely on leader sts to determine what templateHash to use. That is why `createControllerRevisionIfNonExist` checks if the leaderSts exists before attempting to create a new one. When a leaderSts doesn't exist however, we need to determine what templateHash to use. But the templateHash depends on a revision being created. 

So in the case where the leaderSts doesn't exist yet, we would need to set the value of `templateHash` to the hashed controller revision, basically having two labels with the same value. 

Or, we can change the `controller.kubernetes.io/hash` label to be our selector instead. In the case where we pass a templateHash, we set that label equal to the templateHash value.

ahg-g commented on pkg/controllers/leaderworkerset_controller.go:
```diff
@@ -387,6 +410,9 @@ func (r *LeaderWorkerSetReconciler) updateConditions(ctx context.Context, lws *l
 		conditions = append(conditions, makeCondition(leaderworkerset.LeaderWorkerSetUpgradeInProgress))
 	} else if updatedAndReadyCount == int(*lws.Spec.Replicas) {
 		conditions = append(conditions, makeCondition(leaderworkerset.LeaderWorkerSetAvailable))
+		if err := revisionutils.TruncateHistory(ctx, r.Client, lws, templateHash); err != nil {
```
as a follow up, we can store the current and updated revisions for debugging purposes. The current one can be set here, while the updated one can be set in the if block above.

ahg-g commented on pkg/controllers/pod_controller.go:
```diff
@@ -118,8 +119,12 @@ func (r *PodReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.R
 	}
-
-	statefulSet, err := constructWorkerStatefulSetApplyConfiguration(pod, leaderWorkerSet)
+	currentRevision, err := revisionutils.GetLeaderWorkerSetRevisionFromTemplateHash(ctx, r.Client, &leaderWorkerSet, pod.Labels[leaderworkerset.TemplateRevisionHashKey])
```
nit, I would not call it current because it could be current or the updated one, call it revision.

Edwinhr716 commented on pkg/history/controller_history.go:
```diff
@@ -0,0 +1,209 @@
+			return nil, fmt.Errorf("controller Revision with same name but different content exists")
+		}
+	}
+	return revision, err
```
In the case where the webhook defaults, the revision itself would be different, since we store the NetworkConfig in `data.Raw` which is what we use to hash the revision,, so `errors.IsAlreadyexists(err)` returns false. There is no case where there is a match though right? Since there should be a 1:1 mapping?

Edwinhr716 commented:
Addressed second round of comments, added fix for #291 and added e2e test for it. Still missing integration tests and unit tests

ahg-g commented on pkg/utils/revision/revision_utils.go:
```diff
@@ -0,0 +1,368 @@
+// parent is of parentKind. The ControllerRevision has labels matching template labels, contains Data equal to data, and
+// has a Revision equal to revision. If the returned error is nil, the returned ControllerRevision is valid. If the
+// returned error is not nil, the returned ControllerRevision is invalid for use.
+func NewControllerRevision(parent metav1.Object,
```
No need for this function on its own, just move this code inside `NewRevision` and let all the tests use `NewRevision`; this means the parentKind is hardcoded and the patch for `Data` is created in one place.

ahg-g commented on pkg/utils/revision/revision_utils.go:
```diff
@@ -0,0 +1,368 @@
+	return nil
+}
+
+// Sha1Hash accepts an input string and returns the 40 character SHA1 hash digest of the input string.
```
do we need this?

Edwinhr716 replied:
Yes, we use this function to create the GroupKey https://github.com/kubernetes-sigs/lws/blob/main/pkg/webhooks/pod_webhook.go#L176

ahg-g replied:
ok, then we need to move it somewhere else, this is not related to revision management.

ahg-g commented on pkg/utils/revision/revision_utils.go:
```diff
@@ -0,0 +1,368 @@
+	return json.Marshal(objCopy)
+}
+
+func CreateLeaderWorkerSetRevision(
```
we can merge this and `CreateControllerRevision` named `CreateRevision`

ahg-g commented on pkg/utils/revision/revision_utils.go:
```diff
@@ -0,0 +1,368 @@
+	return bytes.Equal(lhs.Data.Raw, rhs.Data.Raw) && apiequality.Semantic.DeepEqual(lhs.Data.Object, rhs.Data.Object)
+}
+
+type History struct {
+	client.Client
+	context context.Context
+}
```
We can get rid of this completely.

ahg-g commented on pkg/utils/revision/revision_utils.go:
```diff
@@ -0,0 +1,368 @@
+// ListControllerRevisions lists all ControllerRevisions matching selector and owned by parent or no other
+// controller. If the returned error is nil the returned slice of ControllerRevisions is valid. If the
+// returned error is not nil, the returned slice is not valid.
+func (h *History) ListControllerRevisions(parent metav1.Object, selector labels.Selector) ([]*appsv1.ControllerRevision, error) {
```
Just pass in the context and client and remove the struct. I would also rename it to ListRevisions to match all the other functions in the file.

ahg-g commented on pkg/utils/revision/revision_utils.go:
```diff
@@ -0,0 +1,368 @@
+
+// HashControllerRevision hashes the contents of revision's Data using FNV hashing.
+// The returned hash will be a safe encoded string to avoid bad words.
+func HashControllerRevision(revision *appsv1.ControllerRevision) string {
```
`HashRevision`

ahg-g commented on pkg/utils/revision/revision_utils.go:
```diff
@@ -0,0 +1,368 @@
+
+// ControllerRevisionName returns the Name for a ControllerRevision in the form prefix-hash. If the length
+// of prefix is greater than 223 bytes, it is truncated to allow for a name that is no larger than 253 bytes.
+func ControllerRevisionName(prefix string, hash string) string {
```
`RevisionName`

ahg-g commented on pkg/utils/revision/revision_utils.go:
```diff
@@ -0,0 +1,368 @@
+// CreateControllerRevision attempts to create the revision as owned by parent via a ControllerRef. If the returned
+// error is not nil, creation failed. If the returned error is nil, the returned ControllerRevision has been
+// created.
+func (h *History) CreateControllerRevision(parent metav1.Object, revision *appsv1.ControllerRevision) (*appsv1.ControllerRevision, error) {
```
As mentioned below, please move this code inside `CreateLeaderWorkerSetRevision` and rename the function to `CreateRevision`

ahg-g commented on pkg/utils/revision/revision_utils.go:
```diff
@@ -0,0 +1,368 @@
+// DeleteControllerRevision attempts to delete revision. If the returned error is not nil, deletion has failed.
+func (h *History) DeleteControllerRevision(revision *appsv1.ControllerRevision) error {
+	return h.Delete(h.context, revision)
+}
```
this function is redundant, please remove and just use Delete directly instead

ahg-g commented on pkg/utils/revision/revision_utils.go:
```diff
@@ -0,0 +1,368 @@
+	return []byte(fmt.Sprintf(`{"metadata":{"ownerReferences":[{"$patch":"delete","uid":"%s"}],"uid":"%s"}}`, revisionUID, parentUID))
+}
+
+func GetLeaderWorkerSetRevisionFromTemplateHash(ctx context.Context, k8sClient client.Client, lws *leaderworkerset.LeaderWorkerSet, templateHash string) (*appsv1.ControllerRevision, error) {
```
```suggestion
func GetRevision(ctx context.Context, k8sClient client.Client, lws *leaderworkerset.LeaderWorkerSet, templateHash string) (*appsv1.ControllerRevision, error) {
```

ahg-g commented on pkg/utils/revision/revision_utils_test.go:
```diff
@@ -0,0 +1,120 @@
+	}
+}
+
+func BuildLeaderWorkerSet(nsName string) *leaderworkerset.LeaderWorkerSet {
```
Lets use the wrappers in https://github.com/kubernetes-sigs/lws/blob/main/test/testutils/wrappers.go

Edwinhr716 replied:
Can't, otherwise it creates an import cycle. testutils imports the package revision in order to use GetPatch, and NewControllerRevision. So if I import testutils, the import cycle is testutils -> revision -> testutils. It was either recreating this function, or having a copy of GetPatch and NewController in testutils. Similar problem that we had with `tpu_test.go` https://github.com/kubernetes-sigs/lws/blob/main/pkg/utils/accelerators/tpu_test.go#L247

ahg-g replied:
We need to split the wrappers and validators into separate pkgs then. We can do that as a followup though.

ahg-g commented on pkg/utils/revision/revision_utils.go:
```diff
@@ -0,0 +1,368 @@
+
+// TruncateHistory cleans up all other controller revisions except the currentRevision.
+// currentRevision is the one that matches the templateHash that is passed
+func TruncateHistory(ctx context.Context, k8sClient client.Client, lws *leaderworkerset.LeaderWorkerSet, templateHash string) error {
```
`TruncateRevisions`

ahg-g commented on pkg/utils/revision/revision_utils.go:
```diff
@@ -0,0 +1,368 @@
+	return controllerHistory.CreateControllerRevision(lws, currentRevision)
+}
+
+// newRevision creates a new ControllerRevision containing a patch that reapplies the target state of LeaderWorkerSet.
```
Just so we distinguish it from `CreateRevision` which creates it in the apiserver.

```suggestion
// newRevision instantiates a ControllerRevision object containing a patch that reapplies the target state of LeaderWorkerSet.
```

ahg-g commented on pkg/utils/revision/revision_utils.go:
```diff
@@ -0,0 +1,368 @@
+	}
+}
+
+func GenerateDeleteOwnerRefStrategicMergeBytes(revisionUID types.UID, parentUID types.UID) []byte {
```
do we need this?

ahg-g reviewed (commented):
Mostly more refactoring to the revision pkg, but I think there is one bug

ahg-g commented on pkg/utils/revision/revision_utils.go:
```diff
@@ -0,0 +1,368 @@
+	if err != nil {
+		return nil, err
+	}
+	revisions, err := controllerHistory.ListControllerRevisions(lws, selector)
```
move this inside NewRevision please

Edwinhr716 commented on pkg/controllers/leaderworkerset_controller.go:
```diff
@@ -523,8 +544,54 @@ func (r *LeaderWorkerSetReconciler) iterateReplicas(ctx context.Context, lws *le
+		return revisionutils.CreateLeaderWorkerSetRevision(ctx, r.Client, lws, sts.Labels[leaderworkerset.TemplateRevisionHashKey])
+	}
+
+	return revisionutils.CreateLeaderWorkerSetRevision(ctx, r.Client, lws, "")
```
It is needed if the LWS object has changed. Otherwise on update it will return the "old" revision, and never detect it. If it isn't being updated, it will return `stsRevision` anyway.

ahg-g replied:
The way we are detecting if an update happened is not what I expected, and it is perhaps related to this, see this comment: https://github.com/kubernetes-sigs/lws/pull/277/files#r1898160442

ahg-g commented on pkg/controllers/leaderworkerset_controller.go:
```diff
@@ -523,8 +544,54 @@ func (r *LeaderWorkerSetReconciler) iterateReplicas(ctx context.Context, lws *le
+		return false, nil
+	}
+
+	stsRevision, err := revisionutils.GetLeaderWorkerSetRevisionFromTemplateHash(ctx, r.Client, lws, sts.Labels[leaderworkerset.TemplateRevisionHashKey])
```
The revision that is passed in is the one that matches the templateHash currently on the sts, so here we should be creating a NewRevision in-memory using the internal hash function and compare it with the one being passed in.

This function should also return this revision in case the comparison fails so that we create this updated revision in case it doesn't exist, and so instead of the boolean, we return a pointer to the updated revision if an update happened.

Edwinhr716 replied:
So create a new revision, and call `EqualRevision()` on the new created one, and the sts one? I don't think that solves the bug. The bug that I see in the current implementation is that if the revision has a templateHash created before the CR change, once we create a new one to detect update, it will incorrectly detect it as it will now use the new hashing function. This can be fixed by removing the templateHash check from `equalRevision`. I don't see how this changes fixes that bug.

Edwinhr716 commented on pkg/controllers/leaderworkerset_controller.go:
```diff
@@ -523,8 +544,54 @@ func (r *LeaderWorkerSetReconciler) iterateReplicas(ctx context.Context, lws *le
+}
+
+func (r *LeaderWorkerSetReconciler) createControllerRevisionIfNonExist(ctx context.Context, sts *appsv1.StatefulSet, lws *leaderworkerset.LeaderWorkerSet) (*appsv1.ControllerRevision, error) {
+	if sts == nil {
+		return revisionutils.CreateLeaderWorkerSetRevision(ctx, r.Client, lws, "")
+	}
+
+	stsRevision, err := revisionutils.GetLeaderWorkerSetRevisionFromTemplateHash(ctx, r.Client, lws, sts.Labels[leaderworkerset.TemplateRevisionHashKey])
+	if err != nil {
+		return nil, err
+	}
+
+	if stsRevision == nil {
+		return revisionutils.CreateLeaderWorkerSetRevision(ctx, r.Client, lws, sts.Labels[leaderworkerset.TemplateRevisionHashKey])
+	}
+
+	return revisionutils.CreateLeaderWorkerSetRevision(ctx, r.Client, lws, "")
```
Since this is the only place where we create revisions, what happens when the LWS object changes? At this point, `leaderSts` exists, `stsRevision` exists, but we still need to create a new revision, which we don't here. it would return at line 566

ahg-g replied:
See this comment: https://github.com/kubernetes-sigs/lws/pull/277/files#r1898160442

Edwinhr716 commented on pkg/controllers/leaderworkerset_controller.go:
```diff
@@ -523,8 +544,54 @@ func (r *LeaderWorkerSetReconciler) iterateReplicas(ctx context.Context, lws *le
 }
 
+func (r *LeaderWorkerSetReconciler) getLeaderStatefulSet(ctx context.Context, lws *leaderworkerset.LeaderWorkerSet) (*appsv1.StatefulSet, error) {
+	sts := &appsv1.StatefulSet{}
+	err := r.Get(ctx, types.NamespacedName{Name: lws.Name, Namespace: lws.Namespace}, sts)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	return sts, nil
```
Can't do that, the client expects the pointer to be initialized. Tested it out and got this error ``` "error": "expected pointer, but got nil"} ```

Edwinhr716 commented:
Addressed third round of comments

ahg-g reviewed (commented):
Great, the logic is more intuitive now and easier to reason about.

Reviewed leaderworkerset_controller.go, will continue with the rest, but I thought it would be useful to send those comments early, mostly to polish comments and function names.

Also, what do you think about changing `templateHash` to `revisionKey` everywhere in the code apart from the label name? This makes it clearer what it represents.

ahg-g commented on pkg/controllers/pod_controller.go:
```diff
@@ -180,6 +185,9 @@ func (r *PodReconciler) handleRestartPolicy(ctx context.Context, pod corev1.Pod,
 		if err := r.Get(ctx, types.NamespacedName{Name: leaderPodName, Namespace: pod.Namespace}, &leader); err != nil {
 			return false, err
 		}
+		if leader.Labels[leaderworkerset.TemplateRevisionHashKey] != pod.Labels[leaderworkerset.TemplateRevisionHashKey] {
```
Add a comment explaining why we do this: basically if the revision is different from the leader, then we expect this pod will be deleted soon and an alternative will be created using the new revision.

ahg-g commented on pkg/utils/revision/revision_utils.go:
```diff
@@ -0,0 +1,301 @@
+// Functions in this package are adapted from https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/statefulset/ and
+// https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/history/controller_history.go
+
+// ControllerRevisionName returns the Name for a ControllerRevision in the form prefix-hash-revisionnumber. If the length
```
update the comments pls with the new function names.

ahg-g commented on pkg/utils/revision/revision_utils.go:
```diff
@@ -0,0 +1,301 @@
+// revision-number allows us to avoid collisions if the created prefix-hash already exists in the history, since revision
+// will be unique.
+func RevisionName(prefix string, hash string, revisionNumber int64) string {
+	if len(prefix) > 223 {
+		prefix = prefix[:223]
```
We need to use a shorter prefix now that we have the revision number as part of the name.

```suggestion
	if len(prefix) > 220 {
		prefix = prefix[:220]
```

ahg-g commented on pkg/utils/revision/revision_utils.go:
```diff
@@ -0,0 +1,301 @@
+// EqualRevision returns true if lhs and rhs are either both nil, if the templateRevisionHash is the same,
+// or if they are semantically equivalent.
+func EqualRevision(lhs *appsv1.ControllerRevision, rhs *appsv1.ControllerRevision) bool {
+
```
nit: remove empty line please

ahg-g commented on pkg/utils/revision/revision_utils.go:
```diff
@@ -0,0 +1,301 @@
+		return lhs == rhs
+	}
+
+	if lhs.Labels[leaderworkerset.TemplateRevisionHashKey] == rhs.Labels[leaderworkerset.TemplateRevisionHashKey] {
```
check that both labels are not nil

ahg-g commented on pkg/utils/revision/revision_utils.go:
```diff
@@ -0,0 +1,301 @@
+	spec := raw["spec"].(map[string]interface{})
+	specCopy["networkConfig"] = spec["networkConfig"]
+	specCopy["leaderWorkerTemplate"] = spec["leaderWorkerTemplate"].(map[string]interface{})
+	specCopy["$patch"] = "replace"
```
I think this will cause the whole spec to be replaced, what we want is to replace the `leaderWorkerTemplate` and `networkConfig` fields only, the rest should stay the same.

The unit test can catch this: try to set a field in the incoming LWS spec other than networkConfig and leaderWorkerTemplate, the restored lws should have the field, I think now it would not.

Edwinhr716 replied:
Yes, it does replace the whole spec, sets the other spec fields to nil. Fixed and changed unit test to account for it.

ahg-g commented on pkg/utils/revision/revision_utils.go:
```diff
@@ -0,0 +1,301 @@
+
+// HashControllerRevision hashes the contents of revision's Data using FNV hashing.
+// The returned hash will be a safe encoded string to avoid bad words.
+func HashRevision(revision *appsv1.ControllerRevision) string {
```
can we move this and the functions related to hashing to the bottom beside each other and make them private (start with lower case ) please; I believe those functions should not be used outside the pkg

ahg-g commented on pkg/utils/revision/revision_utils.go:
```diff
@@ -0,0 +1,301 @@
+// of prefix is greater than 223 bytes, it is truncated to allow for a name that is no larger than 253 bytes.
+// revision-number allows us to avoid collisions if the created prefix-hash already exists in the history, since revision
+// will be unique.
+func RevisionName(prefix string, hash string, revisionNumber int64) string {
```
This should be private to the pkg

ahg-g commented on pkg/utils/revision/revision_utils.go:
```diff
@@ -0,0 +1,301 @@
+	return owned, err
+}
+
+func DeepHashObject(hasher hash.Hash, objectToWrite interface{}) {
```
private function

ahg-g commented on pkg/utils/revision/revision_utils.go:
```diff
@@ -0,0 +1,301 @@
+// previous version. If the returned error is nil the patch is valid. The current state that we save is the
+// leaderWorkerTemplate and NetworkConfig. We can modify this later to encompass more state (or less) and
+// remain compatible with previously recorded patches.
+func GetPatch(lws *leaderworkerset.LeaderWorkerSet) ([]byte, error) {
```
private function

ahg-g commented on pkg/utils/revision/revision_utils.go:
```diff
@@ -0,0 +1,301 @@
+// nextRevision finds the next valid revision number based on revisions. If the length of revisions
+// is 0 this is 1. Otherwise, it is 1 greater than the largest revision's Revision. This method
+// assumes that revisions has been sorted by Revision.
+func NextRevision(revisions []*appsv1.ControllerRevision) int64 {
```
private function

ahg-g reviewed (commented):
Reviewed the pod_controller and the revision pkg; in the revision pkg, the main comment is to make some functions private to the pkg and group them at the end of the file please so that it is clear what functions are supposed to be used externally.

Edwinhr716 commented on pkg/controllers/leaderworkerset_controller.go:
```diff
@@ -96,13 +100,40 @@ func (r *LeaderWorkerSetReconciler) Reconcile(ctx context.Context, req ctrl.Requ
+	updatedRevision, err := r.leaderWorkerSetUpdated(ctx, leaderSts, lws, revision)
+	lwsUpdated := updatedRevision != nil
+	if lwsUpdated {
+		revision, err = revisionutils.CreateRevision(ctx, r.Client, lws, updatedRevision.Labels[leaderworkerset.TemplateRevisionHashKey])
```
Makes sense, we can use `revision.Namespace` to extract the namespace for the Get call as well

Edwinhr716 commented on pkg/controllers/leaderworkerset_controller.go:
```diff
@@ -101,6 +101,7 @@ func (r *LeaderWorkerSetReconciler) Reconcile(ctx context.Context, req ctrl.Requ
 	ctx = ctrl.LoggerInto(ctx, log)
 
 	leaderSts, err := r.getLeaderStatefulSet(ctx, lws)
+	log.V(2).Info(fmt.Sprintf("leader sts %v", leaderSts))
```
Was debugging, forgot to remove.I'll just delete it

ahg-g commented on pkg/controllers/leaderworkerset_controller.go:
```diff
@@ -523,8 +553,48 @@ func (r *LeaderWorkerSetReconciler) iterateReplicas(ctx context.Context, lws *le
+	return sts, nil
+}
+
+func (r *LeaderWorkerSetReconciler) createControllerRevisionIfNonExist(ctx context.Context, sts *appsv1.StatefulSet, lws *leaderworkerset.LeaderWorkerSet) (*appsv1.ControllerRevision, error) {
```
actually this should be "IfNotExists" not "IfNoneExists" because we may create even if one exists that doesn't match the revisionKey we have.

ahg-g commented on pkg/utils/revision/revision_utils.go:
```diff
@@ -0,0 +1,301 @@
+	leaderworkerset "sigs.k8s.io/lws/api/leaderworkerset/v1"
+)
+
+// Functions in this package are adapted from https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/statefulset/ and
```
I would create a GetRevisionKey func as follows and use it instead of access the label directly in many places; this also allows us to check for nil label in one place:

```
func GetRevisionKey(obj metav1.Object) string {
  if obj.Labels != nil {
    return obj.Labels[leaderworkerset.TemplateRevisionHashKey]
  }
  return ""
}
```

ahg-g replied:
this function can be used by pod, sts and revision objects

Edwinhr716 commented on pkg/utils/revision/revision_utils.go:
```diff
@@ -0,0 +1,301 @@
+	if len(revisions) > 1 {
+		// Since we only create a controllerRevision when the template hash changes, only one should match
+		log.Error(err, "More than one revision exists for the given template hash; returning the latest revision")
+		return revisions[len(revisions)-1], nil
```
@ahg-g is there a guarantee that the latest is the one it the highest index? The revisions are not sorted

ahg-g replied:
no, we need to sort here.

Edwinhr716 replied:
Instead of sorting, wdyt of iterating through the list and returning the one with highest revision value?

ahg-g replied:
oh, yeah, if that is all we need

ahg-g replied:
change the semantics of `nextRevision` to `getHighestRevision`, use it here and in NewRevision (where you would increment by one to get the next revision

Edwinhr716 commented:
Finished addressing comments. I agree, makes sense to change `templateHash` to `revisionKey`. Also changed that in the latest commit.

ahg-g commented on pkg/controllers/pod_controller.go:
```diff
@@ -119,7 +119,7 @@ func (r *PodReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.R
 		return ctrl.Result{}, nil
 	}
-	revision, err := revisionutils.GetRevision(ctx, r.Client, &leaderWorkerSet, pod.Labels[leaderworkerset.TemplateRevisionHashKey])
+	revision, err := revisionutils.GetRevision(ctx, r.Client, &leaderWorkerSet, pod.Labels[leaderworkerset.RevisionKey])
```
```suggestion
	revision, err := revisionutils.GetRevision(ctx, r.Client, &leaderWorkerSet, revisionutils.GetRevisionKey(pod))
```

ahg-g commented on pkg/controllers/pod_controller.go:
```diff
@@ -185,7 +185,8 @@ func (r *PodReconciler) handleRestartPolicy(ctx context.Context, pod corev1.Pod,
 		}
-		if leader.Labels[leaderworkerset.TemplateRevisionHashKey] != pod.Labels[leaderworkerset.TemplateRevisionHashKey] {
+		// Different revision key means that this pod will be deleted soon and alternative will be created with the matching key
+		if leader.Labels[leaderworkerset.RevisionKey] != pod.Labels[leaderworkerset.RevisionKey] {
```
```suggestion
		if revisionutils.GetRevisionKey(leader) != revisionutils.GetRevisionKey(pod) {
```

ahg-g commented on pkg/controllers/pod_controller.go:
```diff
@@ -292,7 +293,7 @@ func constructWorkerStatefulSetApplyConfiguration(leaderPod corev1.Pod, lws lead
 		leaderworkerset.SetNameLabelKey:         lws.Name,
 		leaderworkerset.GroupUniqueHashLabelKey: leaderPod.Labels[leaderworkerset.GroupUniqueHashLabelKey],
-		leaderworkerset.TemplateRevisionHashKey: leaderPod.Labels[leaderworkerset.TemplateRevisionHashKey],
+		leaderworkerset.RevisionKey:             leaderPod.Labels[leaderworkerset.RevisionKey],
```
```suggestion
		leaderworkerset.RevisionKey:             revisionutils.GetRevisionKey(leaderPod),
```

ahg-g commented on pkg/utils/revision/revision_utils.go:
```diff
@@ -291,11 +198,105 @@ func TruncateRevisions(ctx context.Context, k8sClient client.Client, lws *leader
 
 	for i, revision := range revisions {
-		if revision.Labels[leaderworkerset.TemplateRevisionHashKey] != templateHash {
+		if revision.Labels[leaderworkerset.RevisionKey] != revisionKey {
```
```suggestion
		if GetRevisionKey(revision) != revisionKey {
```

ahg-g commented on test/testutils/util.go:
```diff
@@ -54,10 +54,10 @@ func CreateWorkerPodsForLeaderPod(ctx context.Context, leaderPod corev1.Pod, k8s
+						leaderworkerset.SetNameLabelKey:     lws.Name,
+						"worker.pod":                        "workers",
+						leaderworkerset.WorkerIndexLabelKey: strconv.Itoa(i),
+						leaderworkerset.RevisionKey:         leaderPod.Labels[leaderworkerset.RevisionKey],
```
```suggestion
						leaderworkerset.RevisionKey:         revisionutils.GetRevisionKey(leaderPod),
```

ahg-g commented on test/testutils/util.go:
```diff
@@ -229,8 +229,8 @@ func GetLeaderPod(ctx context.Context, lws *leaderworkerset.LeaderWorkerSet, k8s
 		}
-		if cr.Labels[leaderworkerset.TemplateRevisionHashKey] != pod.Labels[leaderworkerset.TemplateRevisionHashKey] {
-			return fmt.Errorf("TemplateHash does not match, expected %s, got %s", cr.Labels[leaderworkerset.TemplateRevisionHashKey], pod.Labels[leaderworkerset.TemplateRevisionHashKey])
+		if revisionutils.GetRevisionKey(cr) != pod.Labels[leaderworkerset.RevisionKey] {
```
```suggestion
		if revisionutils.GetRevisionKey(cr) != revisionutils.GetRevisionKey(pod) {
```

ahg-g commented on test/testutils/util.go:
```diff
@@ -229,8 +229,8 @@ func GetLeaderPod(ctx context.Context, lws *leaderworkerset.LeaderWorkerSet, k8s
-		if cr.Labels[leaderworkerset.TemplateRevisionHashKey] != pod.Labels[leaderworkerset.TemplateRevisionHashKey] {
-			return fmt.Errorf("TemplateHash does not match, expected %s, got %s", cr.Labels[leaderworkerset.TemplateRevisionHashKey], pod.Labels[leaderworkerset.TemplateRevisionHashKey])
+		if revisionutils.GetRevisionKey(cr) != pod.Labels[leaderworkerset.RevisionKey] {
+			return fmt.Errorf("TemplateHash does not match, expected %s, got %s", revisionutils.GetRevisionKey(cr), pod.Labels[leaderworkerset.RevisionKey])
```
```suggestion
			return fmt.Errorf("TemplateHash does not match, expected %s, got %s", revisionutils.GetRevisionKey(cr), revisionutils.GetRevisionKey(pod))
```

ahg-g commented on test/testutils/util.go:
```diff
@@ -283,7 +283,7 @@ func SetLeaderPodToReady(ctx context.Context, k8sClient client.Client, podName s
 		}
 
-		leaderPod.Labels[leaderworkerset.TemplateRevisionHashKey] = cr.Labels[leaderworkerset.TemplateRevisionHashKey]
+		leaderPod.Labels[leaderworkerset.RevisionKey] = revisionutils.GetRevisionKey(cr)
```
```suggestion
		revisionutils.GetRevisionKey(leaderPod) = revisionutils.GetRevisionKey(cr)
```

Edwinhr716 replied:
Doesn't make sense to use it here since we are setting the field, not retrieving the value

ahg-g replied:
yup, I was doing a sweep on all places where we use `leaderworkerset.RevisionKey`

ahg-g commented on test/testutils/validators.go:
```diff
@@ -279,9 +279,9 @@ func ExpectValidWorkerStatefulSets(ctx context.Context, leaderWorkerSet *leaderw
-			if sts.Labels[leaderworkerset.TemplateRevisionHashKey] != hash {
-				return fmt.Errorf("mismatch template revision hash for worker statefulset, got: %s, want: %s", sts.Labels[leaderworkerset.TemplateRevisionHashKey], hash)
+			hash := revisionutils.GetRevisionKey(cr)
+			if sts.Labels[leaderworkerset.RevisionKey] != hash {
```
```suggestion
			if revisionutils.GetRevisionKey(sts) != hash {
```

ahg-g commented on pkg/controllers/leaderworkerset_controller.go:
```diff
@@ -379,7 +386,7 @@ func (r *LeaderWorkerSetReconciler) updateConditions(ctx context.Context, lws *l
 			readyCount++
 		}
-		if (noWorkerSts || sts.Labels[leaderworkerset.TemplateRevisionHashKey] == templateHash) && pod.Labels[leaderworkerset.TemplateRevisionHashKey] == templateHash {
+		if (noWorkerSts || sts.Labels[leaderworkerset.RevisionKey] == revisionKey) && pod.Labels[leaderworkerset.RevisionKey] == revisionKey {
```
```suggestion
		if (noWorkerSts || revisionutils.GetRevisionKey(sts) == revisionKey) && revisionutils.GetRevisionKey(pod) == revisionKey {
```

ahg-g commented on pkg/controllers/leaderworkerset_controller.go:
```diff
@@ -522,17 +527,17 @@ func (r *LeaderWorkerSetReconciler) iterateReplicas(ctx context.Context, lws *le
 
-		podTemplateHash := sortedPods[index].Labels[leaderworkerset.TemplateRevisionHashKey]
-		if !(podTemplateHash == templateHash && podutils.PodRunningAndReady(sortedPods[index])) {
+		podTemplateHash := sortedPods[index].Labels[leaderworkerset.RevisionKey]
```
```suggestion
		podTemplateHash := revisionutils.GetRevisionKey(sortedPods[index])
```

ahg-g commented on pkg/controllers/leaderworkerset_controller.go:
```diff
@@ -522,17 +527,17 @@ func (r *LeaderWorkerSetReconciler) iterateReplicas(ctx context.Context, lws *le
 
-		stsTemplateHash := sortedSts[index].Labels[leaderworkerset.TemplateRevisionHashKey]
-		return stsTemplateHash == templateHash && statefulsetutils.StatefulsetReady(sortedSts[index])
+		stsTemplateHash := sortedSts[index].Labels[leaderworkerset.RevisionKey]
+		return stsTemplateHash == revisionKey && statefulsetutils.StatefulsetReady(sortedSts[index])
```
```suggestion
		stsRevisionKey := revisionutils.GetRevisionKey(sortedSts[index])
		return stsRevisionKey == revisionKey && statefulsetutils.StatefulsetReady(sortedSts[index])
```

ahg-g commented on pkg/utils/revision/revision_utils.go:
```diff
@@ -0,0 +1,302 @@
+
+	leaderworkerset "sigs.k8s.io/lws/api/leaderworkerset/v1"
+)
+
```
One more nit, typically we would organize the functions in a way similar to how they are typically used; so I would reorganize them as follows:

- NewRevision
- CreateRevision
- GetRevisionKey
- GetRevision
- ListRevisions
- ApplyRevision
- EqualRevision
- TruncateRevisions

ahg-g commented on pkg/utils/revision/revision_utils.go:
```diff
@@ -0,0 +1,302 @@
+		return nil, err
+	}
+
+	templateLabels := map[string]string{
+		leaderworkerset.RevisionKey:     revisionKey,
+		leaderworkerset.SetNameLabelKey: lws.Name,
+	}
+
+	cr := &appsv1.ControllerRevision{
+		ObjectMeta: metav1.ObjectMeta{
+			Labels:          templateLabels,
+			OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(lws, controllerKind)},
+			Namespace:       lws.Namespace,
+		},
+		Data:     runtime.RawExtension{Raw: patch},
+		Revision: revision,
+	}
+
+	hash := hashRevision(cr)
+	cr.Name = revisionName(lws.Name, hash, revision)
+	if cr.Labels[leaderworkerset.RevisionKey] == "" {
+		cr.Labels[leaderworkerset.RevisionKey] = hash
+	}
```
```suggestion
	hash := hashRevision(cr)	
	if revisionKey == "" {
		revisionKey = hash
	}
	cr := &appsv1.ControllerRevision{
		ObjectMeta: metav1.ObjectMeta{
			Name: revisionName(lws.Name, hash, revision),
			Namespace:       lws.Namespace,
			Labels:          map[string]string{
		            leaderworkerset.RevisionKey:     revisionKey,
		            leaderworkerset.SetNameLabelKey: lws.Name,
	                 },
			OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(lws, controllerKind)},
		},
		Data:     runtime.RawExtension{Raw: patch},
		Revision: revision,
	}

```

Edwinhr716 replied:
We need to define the variable `cr` before we can calculate the hash. At line 143, `cr` still hasn't been defined yet.

ahg-g replied:
ah, right; perhaps we can still cleanup the label setting a bit

```
cr := &appsv1.ControllerRevision{
		ObjectMeta: metav1.ObjectMeta{
			Namespace:       lws.Namespace,
			Labels:          map[string]string{
		            leaderworkerset.SetNameLabelKey: lws.Name,
	                 },
			OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(lws, controllerKind)},
		},
		Data:     runtime.RawExtension{Raw: patch},
		Revision: revision,
	}
	hash := hashRevision(cr)	
	if revisionKey == "" {
		revisionKey = hash
	}
        cr.Name = revisionName(lws.Name, hash, revision)
        cr.Labels[leaderworkerset.RevisionKey] = revisionKey
```

Edwinhr716 commented:
Addressed comments. Flakiness of tests is much higher now that everything uses `GetRevisionKey()`, unsure why. I'll retest later to see if that fixes it. 

FYI I do want to add more integration tests before merging which is why I had the WIP tag in the title

ahg-g commented on test/e2e/e2e_test.go:
```diff
@@ -301,4 +303,60 @@ var _ = ginkgo.Describe("leaderWorkerSet e2e tests", func() {
+				return fmt.Errorf("StatefulSet did not have the expected container name")
+			}
+			return nil
+		}, timeout, interval).Should(gomega.Succeed())
```
This test will be flaky, how do you guarantee that the update isn't faster than your check here?

Edwinhr716 replied:
There is no guarantee, that is just a risk that is taken when checking the state mid update. Similar issue in other e2e tests. All I can do is have as many replicas as possible so that the update takes long to reach the "-0" leader sts.

ahg-g replied:
We can't have tests that don't offer guarantees, those will flake, which means when they fail, we don't know if they fail because of a bug or the race condition. The test you are trying to do here should be done via an integration test  where we control update progress.

Edwinhr716 replied:
Right, but other update tests also don't offer guarantees. For example this one https://github.com/kubernetes-sigs/lws/blob/main/test/e2e/e2e_test.go#L181, what guarantees that the update isn't done before this check is performed? I guess I'm just confused as to why we can do it there but not here, both are checking the state mid update.

ahg-g commented on pkg/controllers/leaderworkerset_controller.go:
```diff
@@ -523,8 +558,54 @@ func (r *LeaderWorkerSetReconciler) iterateReplicas(ctx context.Context, lws *le
+
+func (r *LeaderWorkerSetReconciler) getOrCreateRevisionIfNonExist(ctx context.Context, sts *appsv1.StatefulSet, lws *leaderworkerset.LeaderWorkerSet) (*appsv1.ControllerRevision, error) {
+	revisionKey := ""
+	if sts != nil {
+		// Uses the hash in the leader sts to avoid detecting update in the case where LWS controller is upgraded from a version where
+		// the revisionKey was used to detect update instead of controller revision.
+		revisionKey = revisionutils.GetRevisionKey(sts)
+	}
```
```suggestion
	// Uses the hash in the leader sts to avoid detecting update in the case where LWS controller is upgraded from a version where
	// the revisionKey was used to detect update instead of controller revision.
	revisionKey := revisionutils.GetRevisionKey(sts)
```

ahg-g replied:
GetRevisionKey already does a nil check

Edwinhr716 commented on test/e2e/e2e_test.go:
```diff
@@ -301,4 +303,60 @@ var _ = ginkgo.Describe("leaderWorkerSet e2e tests", func() {
+		testing.ExpectLeaderWorkerSetAvailable(ctx, k8sClient, lws, "All replicas are ready")
+
+	})
+	ginkgo.It("Rolling update with restart policy RecreateGroupOnPodRestart only restarts the leader once", func() {
```
Yes, tested it out before the adding the fix.

Edwinhr716 commented on pkg/controllers/pod_controller_test.go:
```diff
@@ -263,11 +287,85 @@ func TestConstructWorkerStatefulSetApplyConfiguration(t *testing.T) {
 		},
+		{
+			name:     "revision is before update, will use that and the old templateHash to create the worker statefulset configuration",
+			revision: currentRevision,
```
Agreed, it isn't different. Kept it from when it was still an if statement checking the templateHash. I'll remove it

ahg-g commented on test/e2e/e2e_test.go:
```diff
@@ -301,4 +303,60 @@ var _ = ginkgo.Describe("leaderWorkerSet e2e tests", func() {
+
+		gomega.Eventually(func() (bool, error) {
+			return initialLeaderPod.UID == midUpdateLeaderPod.UID, nil
+		}, timeout, interval).Should(gomega.Equal(false))
```
why is this in an eventually clause? nothing is changing inside it

ahg-g commented:
/lgtm
/approve

Will use https://github.com/kubernetes-sigs/lws/issues/294 to track integration tests
